homeassistant/helpers/translation.py

```diff
@@ -91,40 +94,43 @@ def build_resources(
         else:
             domain = component.split(".", 1)[0]
 
-        if domain not in resources:
-            resources[domain] = {}
+        domain_resources = resources.setdefault(domain, {})
 
         # Add the translations for this component to the domain resources.
         # Since clients cannot determine which platform an entity belongs to,
         # all translations for a domain will be returned together.
-        resources[domain].update(translation_cache[component])
+
+        if category is None:
+            domain_resources.update(translation_cache[component])
+            continue
+
+        if category not in translation_cache[component]:
+            continue
+
+        domain_resources.setdefault(category, {}).update(
+            translation_cache[component][category]
+        )
 
     return resources
 
 
-@bind_hass
-async def async_get_component_resources(
-    hass: HomeAssistantType, language: str
+async def async_get_component_cache(
+    hass: HomeAssistantType, language: str, components: Set[str]
 ) -> Dict[str, Any]:
-    """Return translation resources for all components.
-
-    We go through all loaded components and platforms:
-     - see if they have already been loaded (exist in translation_cache)
-     - load them if they have not been loaded yet
-     - write them to cache
-     - flatten the cache and return
-    """
+    """Return translation cache that includes all specified components."""
     # Get cache for this language
-    cache = hass.data.setdefault(TRANSLATION_STRING_CACHE, {})
-    translation_cache = cache.setdefault(language, {})
-
-    # Get the set of components to check
-    components = hass.config.components | await async_get_config_flows(hass)
+    cache: Dict[str, Dict[str, Any]] = hass.data.setdefault(
+        TRANSLATION_STRING_CACHE, {}
+    )
+    translation_cache: Dict[str, Any] = cache.setdefault(language, {})
```

Comment:
Here we seem to set a language as top level key in the translation cache. How does that match with your description in the comment?

Reply:
Sorry, I got confused with all the nested dicts. The translation cache here and what we return is one level below the language key.

Reply:
We return `translation_cache`, which has the described structure.

You're right that `cache` is `{ language: translation_cache }`.